tests/BlazingSQLTest/EndToEndTests/concurrentTest.py

```diff
@@ -0,0 +1,180 @@
+
+
+def datasources(data_types, dask_client, nRals):
+    for fileSchemaType in data_types:
+        if skip_test(dask_client, nRals, fileSchemaType, queryType):
+            continue
+        yield fileSchemaType
+
+
+def samples(data_types, dask_client, nRals):
+    for fileSchemaType in datasources(data_types, dask_client, nRals):
+        i = 0
+        for query in queries:
+            i = i + 1
+            istr = str(i) if i > 10 else "0"+str(i)
+            queryId = "TEST_" + istr
+            sampleId = str(fileSchemaType) + "." + queryId
+            yield sampleId, query, queryId, fileSchemaType
+
+
+def create_tables(bc, data_types, dask_client, nRals, dir_data_lc, tables):
+    print("######## Creating tables queries ...########")
+    for fileSchemaType in datasources(data_types, dask_client, nRals):
+        cs.create_tables(bc, dir_data_lc, fileSchemaType, tables=tables)
+
+
+def start_queries(bc, data_types, dask_client, nRals, tokens):
+    print("######## Starting queries ...########")
+    done = {} # sampleId -> True|False (fetch completed?)
+    for sampleId, query, _, _ in samples(data_types, dask_client, nRals):
+        start_query(bc, tokens, query, sampleId)
+        done[sampleId] = False
+    return done
+
+
+def fetch_results(bc, data_types, dask_client, nRals, tokens, drill, done):
+    print("######## Fetching retults ...########")
+    partial_done = dict((fileSchemaType, 0) for fileSchemaType in data_types)
+    done_count = 0
+    break_flag = False
+    total_samples = len(done)
+    while done_count < total_samples and not break_flag:
+        for sampleId, query, queryId, fileSchemaType in samples(data_types, dask_client, nRals):
+            if not done[sampleId]:
+                if fetch_result(bc, tokens, sampleId, drill, query, queryId, fileSchemaType):
+                    done[sampleId] = True
+                    done_count = done_count + 1
+                    partial_done[fileSchemaType] = partial_done[fileSchemaType] + 1
+                    if len(partial_done) == len(queries):
+                        if Settings.execution_mode == ExecutionMode.GENERATOR:
+                            print("==============================")
+                            break_flag = True
+                            break
+
+
+def executionTest(dask_client, drill, dir_data_lc, bc, nRals):
+    tokens = {} # sampleId(fileSchemaType+"."+queryId) -> token
+
+    # Read Data TPCH------------------------------------------------------
+    tables = [
+        "nation",
+        "region",
+        "customer",
+        "lineitem",
+        "orders",
+        "supplier",
+        "part",
+        "partsupp",
+    ]
+    data_types = [
+        DataType.DASK_CUDF,
+        DataType.CSV,
+        DataType.PARQUET,
+    ]  # TODO json
+
+    if Settings.execution_mode == ExecutionMode.GENERATOR:
+        data_types = [DataType.CSV]
+
+    create_tables(bc, data_types, dask_client, nRals, dir_data_lc, tables)
```

Comment:
How does this work? It seems like here you are creating all the tables for all the different data source types, and then running the queries. But in the queries, the table name is the same, so you are not running the queries on the different datasource types. Am i missing something here?

Reply:
the last update fixed this!